gax/src/main/java/com/google/api/gax/batching/FlowControlEventStats.java

```diff
@@ -0,0 +1,122 @@
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching;
+
+import static com.google.api.gax.batching.FlowController.LimitExceededBehavior;
+
+import com.google.api.core.InternalApi;
+import com.google.api.gax.batching.FlowController.FlowControlException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/** Record the statistics of flow control events. */
+@InternalApi("For google-cloud-java client use only")
+public class FlowControlEventStats {
+
+  // Currently we're only interested in the most recent flow control event, but this class can be
+  // expanded to record other stats.
+  private FlowControlEvent lastFlowControlEvent;
+
+  public synchronized void recordFlowControlEvent(FlowControlEvent event) {
+    if (lastFlowControlEvent == null || event.compareTo(lastFlowControlEvent) > 0) {
+      lastFlowControlEvent = event;
+    }
+  }
+
+  public FlowControlEvent getLastFlowControlEvent() {
+    return lastFlowControlEvent;
+  }
+
+  /**
+   * A flow control event. Record throttled time if {@link LimitExceededBehavior} is {@link
+   * LimitExceededBehavior#Block}, or the exception if the behavior is {@link
+   * LimitExceededBehavior#ThrowException}.
+   */
+  public static final class FlowControlEvent implements Comparable<FlowControlEvent> {
+    private long timestampMs;
+    private Long throttledTimeInMs;
+    private FlowControlException exception;
+
+    public static FlowControlEvent create(long throttledTimeInMs) {
+      return create(System.currentTimeMillis(), throttledTimeInMs);
+    }
+
+    public static FlowControlEvent create(FlowControlException exception) {
+      return create(System.currentTimeMillis(), exception);
+    }
+
+    /** Package-private for use in testing. */
+    @VisibleForTesting
+    static FlowControlEvent create(long timestampMs, long throttledTimeInMs) {
```

Comment:
Please use different names for the 2 create methods to make it more obvious what they are used for. Something along the lines of
createReserveDelayed(timestamp, timeWaited)
createReserveDenied(timestamp, exception)